python/taichi/lang/misc.py

```diff
@@ -705,6 +707,85 @@ def Tape(loss, clear_gradients=True):
     return impl.get_runtime().get_tape(loss)
 
 
+def fwdAD(loss,
+          parameters,
+          seed=None,
+          keep_primal=True,
+          clear_gradients=True,
+          clear_loss=True):
+    impl.get_runtime().materialize()
+    if not isinstance(loss, list):
+        loss = [loss]
+
+    # Currently we only support only one N-D field as a group of parameters,
+    # which is sufficient for computing Jacobian-vector product(Jvp).
+    # For cases with multiple groups of parameters, it requires to run the forward ad multiple times,
+    # which is out of scope of the current design for this interface.
+
+    # TODO: support vector field and matrix field
+    assert isinstance(parameters, ScalarField)
+
+    all_fields = [*loss, parameters]
+    fields_without_dual = []
+
+    for x in all_fields:
+        if not x.snode.ptr.has_dual():
+            fields_without_dual.append(x)
+
+    if len(fields_without_dual) > 0:
+        dual_root = FieldsBuilder()
+        for x in fields_without_dual:
+            allocate_dual(x, dual_root)
+        dual_root.finalize()
+
+    def shape_flatten(shape):
+        return functools.reduce((lambda x, y: x * y), list(shape))
+
+    parameters_shape_flatten = shape_flatten(parameters.shape)
+
+    # Handle 0-D field
+    if parameters_shape_flatten == 0:
+        parameters_shape_flatten = 1
+
+    if not seed:
+        # Compute the derivative respect to the first variable by default
+        seed = [0.0 for _ in range(parameters_shape_flatten)]
+        seed[0] = 1.0
+    else:
+        assert parameters_shape_flatten == len(seed)
+
+    # Set seed for each variable
+    if len(seed) == 1:
+        parameters.dual[None] = 1.0 * seed[0]
+    else:
+        for idx, s in enumerate(seed):
+            parameters.dual[idx] = 1.0 * s
+
+    # Clear gradients
+    if clear_gradients:
+        for ls in loss:
+            ls.dual.fill(0)
+
+    # Clear losses
+    if clear_loss:
+        for ls in loss:
+            ls.fill(0)
+
+    return impl.get_runtime().get_fwd_mode_manager(keep_primal)
```

Comment:
Hmmm it feels a bit weird to return a context manager here... I'm happy to discuss this - there might be a better way to organize it.